src/library/dao/playlistdao.cpp

```diff
@@ -980,18 +981,19 @@ void PlaylistDAO::shuffleTracks(const int playlistId, const QList<int>& position
         trackPositionIds.insert(trackBPosition, trackAId);
         newPositions.swap(newPositions.indexOf(trackAPosition),
                           newPositions.indexOf(trackBPosition));
-        QString swapQuery = "UPDATE PlaylistTracks SET position=%1 "
-                "WHERE position=%2 AND playlist_id=%3";
-        query.exec(swapQuery.arg(QString::number(-1),
-                                 QString::number(trackAPosition),
-                                 QString::number(playlistId)));
-        query.exec(swapQuery.arg(QString::number(trackAPosition),
-                                 QString::number(trackBPosition),
-                                 QString::number(playlistId)));
-        query.exec(swapQuery.arg(QString::number(trackBPosition),
-                                 QString::number(-1),
-                                 QString::number(playlistId)));
 
+        qSwap(tracksPositionRowIds[trackAPosition - 1].position,
+                  tracksPositionRowIds[trackBPosition - 1].position);
```

Comment:
But I check now, some new permutation of the sequence is being generated, I can't say is it 100% random or not, but if we assume that the trackAPosition and trackBPosition is random, then i can say that the sequence can't be random. So the previous idea as you say commit 874d511 was ok, as it could guarantee the correctness of the algorithm, but it can be assured that before swapping if we can trace the indies in the list for the given trackAPositon and trackBPosition, then after it surely, swapping those indies would result the correct algorithm